### PR TITLE
fix(contrib): keep distance_transform gradients finite

### LINKED ISSUE #4183: distance_transform returns NaN gradients wherever its convolution hits zero (log(0) through nan_to_num)
## Summary

`kornia.contrib.distance_transform` returns **NaN** gradients for any input that has a region far enough from a boundary pixel for the cascade's convolution to return exactly zero — which includes the ordinary case of a sparse binary mask, and every all-zero input. The forward result is correct; only the backward is poisoned, silently, with no warning.

## Reproduction

`main` @ `ee6b713f`, torch 2.9.1, CPU:

```python
import torch
from kornia.contrib import distance_transform

torch.manual_seed(0)
x = (torch.rand(1, 1, 16, 16) > 0.9).float().requires_grad_()
distance_transform(x, kernel_size=3).sum().backward()
print(torch.isnan(x.grad).sum().item(), "of", x.grad.numel())   # 193 of 256
```

| input (float64) | NaN entries in the gradient |
|---|---|
| random binary mask, 16×16, ~10% ones | 175 / 256 |
| one-hot 4×4 (single boundary pixel) | 15 / 16 |
| all-zeros 4×4 | 16 / 16 |

float32 on the same random mask: 193 / 256. All NaN, no Inf.

## Mechanism

`kornia/contrib/distance_transform.py`, both the 2D and the 3D impl:

```python
cdt = filter2d(boundary, kernel, border_type="replicate")
cdt = -h * torch.log(cdt)
cdt = torch.nan_to_num(cdt, nan=0.0, posinf=0.0, neginf=0.0)
```

Where the convolution returns exactly `0`, `log` gives `-inf`, and `nan_to_num` cleans the *forward* value to `0`. Its backward, however, multiplies the incoming gradient by `isfinite(input)`, so the `inf` derivative of `log` at zero meets a zero multiplier and yields `0 * inf = NaN`. Isolated:

```python
x = torch.zeros(1, 1, 4, 4, dtype=torch.float64, requires_grad=True)
cdt = filter2d(x, torch.ones(1, 3, 3, dtype=torch.float64), border_type="replicate")
torch.nan_to_num(-0.35 * torch.log(cdt), nan=0.0, posinf=0.0, neginf=0.0).sum().backward()
print(x.grad.flatten()[:3])   # tensor([nan, nan, nan], dtype=torch.float64)
```

## Why the existing tests do not catch it

`tests/contrib/test_conv_distance_transformer.py::test_gradcheck` uses `torch.ones(B, C, H, W)` for both the 2D and the 3D case. With an all-ones image the convolution is positive everywhere, the `log` never sees zero, and the defective branch is never entered. The regression test for this needs an input with an actual empty region — a sparse mask or a one-hot image.

## Candidate fix

Guarding the `log` instead of cleaning up after it removes every NaN and leaves the forward **byte-identical**. Measured on the three inputs in the table above, against the library's own output (`torch.equal`, float64):

```python
positive = cdt > 0
cdt = -h * torch.log(torch.where(positive, cdt, torch.ones_like(cdt)))
cdt = torch.where(positive, cdt, torch.zeros_like(cdt))
```

| input | forward byte-identical to current | NaN grads now | NaN grads with the guard |
|---|---|---|---|
| random mask | yes | 193 | **0** |
| one-hot | yes | 15 | **0** |
| all zeros | yes | 16 | **0** |

Note `mask = cdt > 0` is computed *after* this line, on the cleaned value, so a naive `cdt.clamp_min(eps)` before the `log` is **not** a valid alternative: it turns those positions into `-h*log(eps) > 0` and flips them into the mask, changing the forward. Whatever the fix, the forward has to be shown byte-identical.

## Classification

**Broken output** (gradients), reachable from ordinary usage of a differentiable op. Not related to #4152 — verified present and identical on both sides of #4158, which changes nothing here.

## Environment

kornia @ `ee6b713f`, torch 2.9.1, Python 3.11.14, macOS/arm64, CPU.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

### PR DESCRIPTION
**Pull request by Jeevang1-epic**
## What does this pull request do?

Fixes NaN gradients in `kornia.contrib.distance_transform` when an intermediate convolution evaluates to exactly zero.

Previously, both the 2D and 3D implementations evaluated:

```python
cdt = -h * torch.log(cdt)
cdt = torch.nan_to_num(cdt, nan=0.0, posinf=0.0, neginf=0.0)
```

When `cdt == 0`, `torch.log(0)` produces `-inf`. `torch.nan_to_num` repairs the forward value, but the backward still passes through the singular derivative of `log` at zero, which can produce NaN gradients.

The fix avoids evaluating `log` at inactive positions:

```python
positive = cdt > 0
cdt = -h * torch.log(torch.where(positive, cdt, torch.ones_like(cdt)))
cdt = torch.where(positive, cdt, torch.zeros_like(cdt))
```

This is applied to both `_distance_transform_2d_impl` and `_distance_transform_3d_impl`.

The zero-convolution positions therefore follow a finite inactive branch instead of passing through `log(0)`, while positive convolution values retain the existing computation.

The existing `torch.nan_to_num(...)` step is intentionally retained so handling of exceptional non-finite inputs and the surrounding behavior remain unchanged.

The patch does not clamp the convolution result before `log`; doing so would turn zero positions into positive distance values and change the existing mask/forward semantics.

### Regression coverage

A dedicated regression test was added for both:

- 2D input: `(1, 1, 4, 4)`
- 3D input: `(1, 1, 4, 4, 4)`

and for:

- all-zero inputs
- one-hot inputs

across the repository's `float32` / `float64` CPU test matrix.

The regression verifies that:

- the forward output is finite;
- the all-zero case remains exactly zero;
- backward completes successfully;
- `sample.grad` is produced;
- the gradient shape exactly matches the input shape;
- all gradient entries are finite.

The regression was also checked against the unpatched implementation, where the affected cases fail because the gradients contain NaNs.

I additionally compared patched and pre-fix forward results on representative sparse, one-hot, and all-zero inputs using exact `torch.equal` comparisons. The forward results remain unchanged while the affected gradients become finite.

### Compatibility with #4158

This was checked after rebasing onto the current `main`, which already contains #4158.

The two fixes address separate failure modes in the same function:

- #4158 handles kernel working precision / `exp(-dist / h)` underflow;
- this change prevents the later logarithm from being evaluated where the convolution result is zero.

The #4158 behavior, kernel construction, working-dtype handling, `_check_h_range`, offsets, ONNX-safe operations, and public API are unchanged by this patch.

## Related issue or discussion

Fixes #4183.

### Regression test

```text
pixi run test-module tests/contrib/test_conv_distance_transformer.py::TestConvDistanceTransform::test_zero_convolution_has_finite_gradients --device=cpu --dtype=float32,float64

8 passed
```

### Full distance-transform test module

```text
pixi run test-module tests/contrib/test_conv_distance_transformer.py --device=cpu --dtype=float32,float64

42 passed
```

### Lint / formatting

```text
pixi run lint

ruff check    Passed
ruff format   Passed
```

### Type checking

```text
pixi run typecheck
```

Result:

```text
exit code 0
```

The command reports one existing deprecation diagnostic in `kornia/feature/lightglue.py` for `torch.cuda.amp.custom_fwd`; it is unrelated to this change.

### Pre-commit

Targeted pre-commit was run for:

```text
kornia/contrib/distance_transform.py
tests/contrib/test_conv_distance_transformer.py
```

The applicable formatting, lint, merge-conflict, large-file, codespell, and license-header checks pass. On Windows, the mixed-line-ending hook normalized the two touched files when required.

### Git / scope checks

The branch was rebased onto current `upstream/main`.

The final source change is limited to:

```text
kornia/contrib/distance_transform.py
tests/contrib/test_conv_distance_transformer.py
```

with:

```text
2 files changed, 24 insertions(+), 2 deletions(-)
```

`git diff --check` exits successfully.

## How did AI help?

I used AI-assisted tooling to help inspect the existing implementation and test coverage, reproduce the reported backward failure, compare the patch against the issue's requested scope, and organize the validation steps.

I reviewed the resulting source and test changes myself, verified the regression against the pre-fix behavior, checked exact forward-output preservation, confirmed finite gradients after the fix, rebased onto the merged #4158 implementation, and reran the relevant tests, lint, type checking, pre-commit, and Git scope checks before preparing this pull request.

### CODE REVIEW

**Comment by ducha-aiki** [comment]
## Approved at `bb978c08` once a CHANGELOG entry is added — verified on torch 2.9.1 and 2.14.0

Reviewed against merge-base `ca1f2f7b`, CPU.

**Verified**

- **Fail-on-base:** the new `test_zero_convolution_has_finite_gradients` copied onto `main` -> 8 failed on both torch versions (2d/3d × all-zero/one-hot × float32/float64), matching #4183's three reproductions. Head: 42 passed on both versions, and 22 passed under each CI half profile (`cpu-float16`, `cpu-bfloat16`) with no XPASS.
- **Forward is bit-identical:** `torch.equal` to `main` over 36 cases (2D and 3D; random, all-zero, one-hot and all-one inputs; `kernel_size` 3/5/7; `h` 0.35/0.1). Where the convolution is exactly zero the old path produced `-h * log(0) = inf` which `nan_to_num` mapped to `0`; the new path writes `0` directly, and elsewhere the `where` is the identity.
- Gradient: `log` now only ever sees the substituted `1` at inactive positions, so its derivative there is finite and the outer `where` zeroes it; active positions are untouched. Keeping `nan_to_num` afterwards is harmless.

**The one ask:** there is no `CHANGELOG.md` entry. Add a bullet under `## Unreleased` / `### Bug fixes` citing `(#4232)`: `distance_transform` returned NaN gradients wherever the cascade's convolution was exactly zero (sparse masks, all-zero inputs), forward unchanged. `Fixes #4183` is already in the body and registered.

**CI:** the `half-precision (float16)` job on your run hung twice (attempt 1 03:32 -> 09:32 UTC, attempt 2 12:32 -> 18:32, both cancelled at the 6-hour cap, both stalled on entering `tests/feature/test_scale_space_detector.py`, a file this PR does not touch). That is #4214, which hits some Intel runners; the bfloat16 leg and everything else is green. The red `collector` is not yours; it gets re-run after the CHANGELOG push.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Comment by Jeevang1-epic** [comment]
@ducha-aiki  Thanks for the review! I’ve added the requested CHANGELOG entry and updated the branch. All 40 checks are green now. Ready for your review whenever you have time.

**Comment by ducha-aiki** [comment]
Delta re-review at `0eca0d62` — **APPROVED, merge-ready**. The only outstanding ask was the CHANGELOG entry; `0eca0d62` adds it under `### Bug fixes` in the Unreleased section (`+6`, nothing else touched), the code and test files are byte-identical to the approved `bb978c08`, CI is 40/40 green including the float16 leg that hung twice before (#4214), and GitHub reports `MERGEABLE`. Nothing left for you to do.

Posted on behalf of @ducha-aiki by Claude (Claude Fable 5.1).

**Review by ducha-aiki — APPROVED** [review]
